hilti/include/rt/type-info.h

```diff
@@ -0,0 +1,149 @@
+     */
+    const TypeInfo* elementType() const { return _etype; }
+
+    /**
+     * Returns a raw pointer to the element, as well as its type information.
+     */
+    std::pair<const void*, const TypeInfo*> element(const void* p) const { return std::make_pair(_getter(p), _etype); }
+
+private:
+    const TypeInfo* _etype;
+    const Getter _getter;
+};
+
+} // namespace detail
+
+/** Describes a struct field. */
+struct Field {
+    Field(const char* name, const TypeInfo* type, std::ptrdiff_t offset) : name(name), type(type), offset(offset) {}
+
+    std::string name;
+    const TypeInfo* type;
+    std::ptrdiff_t offset;
+};
+
+/** Type-specific information for `bytes`. */
+class Bytes : public detail::AtomicType<hilti::rt::Bytes> {};
+
+/** Type-specific information for signed integers. */
+template<typename T>
+class SignedInteger : public detail::AtomicType<T> {};
+
+/** Type-specific information for unsigned integers. */
+template<typename T>
+class UnsignedInteger : public detail::AtomicType<T> {};
+
+/** Type-specific information for `string`. */
+class String : detail::AtomicType<std::string> {};
+
+/** Type-specific information for `struct`. */
+class Struct {
+public:
+    Struct(std::vector<Field> fields) : _fields(std::move(fields)) {}
+
+    const auto& fields() const { return _fields; }
+
+private:
+    std::vector<Field> _fields;
+};
+
+/** Based class for type-specific information for `value_ref<T>. */
+class ValueReference : public detail::DereferencableType {
+public:
+    using detail::DereferencableType::DereferencableType;
+};
+
+/** Type-specific information for particular instantiations of `value_ref<T>. */
+template<typename T>
+class ValueReferenceFor : public ValueReference {
+public:
+    ValueReferenceFor(const TypeInfo* etype)
+        : ValueReference(etype,
+                         [](const void* p) { return static_cast<const hilti::rt::ValueReference<T>*>(p)->get(); }) {}
+};
+
+} // namespace type_info
+
+// clang-format off
+/**
+ * Top-level information describing one type. There's a generic part
+ * applying to all types, plus a variant storing additional type-specific
+ * information.
+ */
+struct TypeInfo {
+    std::optional<std::string> id; /**< Spicy-side ID associated with the type, if any. */
+    std::string display; /**< String rendering of the type. */
+
+    /**
+     * Type-specific additional information. This also acts as a tag
+     * defining which kind of type is being described.
+     */
+    std::variant<
+        type_info::Bytes,
+        type_info::SignedInteger<int8_t>,
+        type_info::SignedInteger<int16_t>,
+        type_info::SignedInteger<int32_t>,
+        type_info::SignedInteger<int64_t>,
+        type_info::String,
+        type_info::Struct,
+        type_info::ValueReference,
+        type_info::UnsignedInteger<uint8_t>,
+        type_info::UnsignedInteger<uint16_t>,
+        type_info::UnsignedInteger<uint32_t>,
+        type_info::UnsignedInteger<uint64_t>
```

Comment:
Such a closed set of types probably works if we can keep the set of language types quasi-constant on the Spicy-side as well. Else this very strong coupling between the language and users of this API could lead to us breaking users often (as the visitors they'd necessarily implement would not be able to handle all cases absent a common base class).

Once this becomes ready we'll need type info for function-like types as well.

Reply:
Yeah, I imagine the set of types to become very stable. For parameterized types will need only one entry (except for the integers, and that's only to get access to underlying C++ ints).